java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java

```diff
@@ -347,27 +431,45 @@ protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<Z> clazz)
     allInterfaces.addAll(originalInterfaces);
     allInterfaces.addAll(derivedInterfaces.keySet());
     Class<?>[] allInterfacesArray = allInterfaces.toArray(new Class<?>[0]);
+    // ensure we do not hit an existing field
+    int fieldName;
+
+    for (fieldName = 0; fieldName < 8192; fieldName++) {
+      try {
+        clazz.getDeclaredField("___target" + fieldName);
+      } catch (NoSuchFieldException ex) {
+        break;
+      }
+    }
+
+    if (fieldName == 8192) {
+      throw new AssertionError("No free field name found");
+    }
 
     Class<? extends Z> proxy =
         new ByteBuddy()
             .subclass(clazz.isInterface() ? Object.class : clazz)
```

Comment:
this might potentially be a source for a serious memory leak. The reason to that is ByteBuddy always create new instances of "custom" classes, which then get cached by the class loaded for the whole VM lifespan. In order to prevent such issue we must "manually" verify signatures of classes to be created and only construct a new class if its signature differs from already created one. See https://github.com/appium/java-client/blob/42480feca9845bbbb569b13d995b80980bf4ca33/src/main/java/io/appium/java_client/proxy/Helpers.java#L117 for more details

Reply:
This is the root cause for this PR, the cache inside the WebDriverDecorator does prevent from creating alot of classes. But it might be good to make the cache static and weak to allow a complete GC.